daemon/libnetwork/network.go

```diff
@@ -2193,3 +2194,59 @@ func (n *Network) deleteLoadBalancerSandbox() error {
 	}
 	return nil
 }
+
+// State returns the network state, which includes IPAM state.
+func (n *Network) State(ctx context.Context) *network.NetworkState {
+	if n.hasSpecialDriver() {
+		log.G(ctx).Warnf("special driver %s does not support network state", n.Type())
+		return nil
+	}
+
+	ipam, _, err := n.getController().getIPAMDriver(n.ipamType)
+	if err != nil {
+		log.G(ctx).Warnf("failed to get ipam driver %s: %v", n.ipamType, err)
+		return nil
+	}
+
+	var state network.NetworkState
+	state.IPAM = make(map[string]network.IPAMState)
+
+	ipam4Infos, _ := n.IpamInfo()
+	for _, ii := range ipam4Infos {
+		pi, err := defaultipam.PoolIDFromString(ii.PoolID)
+		if err != nil {
+			log.G(ctx).Warnf("failed to parse pool ID %s: %v", ii.PoolID, err)
+			continue
+		}
+		AllocatedIPsInSubnet, AllocatedIPsInPool, err := ipam.GetAllocatedIPs(ii.PoolID)
+		if err != nil {
+			log.G(ctx).Warnf("failed to get available IPs for pool %s: %v", ii.PoolID, err)
+			continue
+		}
+
+		// Calculate the capacity of the pool, which is currently just a CIDR.
+		// Only IPv4 is calculated, as the IPAM state only supports IPv4.
+		capacity := func(cidr netip.Prefix) (uint64, error) {
+			if !cidr.Addr().Is4() {
+				return 0, fmt.Errorf("IPAM state only supports IPv4 CIDRs, got %s", cidr)
+			}
+			return uint64(1) << (32 - cidr.Bits()), nil
```

Comment:
We can do better than hardcode a constant:
```suggestion
			return uint64(1) << (cidr.Addr().BitLen() - cidr.Bits()), nil
```

Reply:
Done